cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java

```diff
@@ -0,0 +1,211 @@
+        throw new IllegalStateException("Expected logger context instance of " + AppenderContext.class.getName() +
+                                          " but got " + context.getClass().getName());
+      }
+
+      fileNamePattern = new FileNamePattern(fileNamePatternStr, this.context);
+    } else {
+      LOG.error(FNP_NOT_SET);
+      throw new IllegalStateException(FNP_NOT_SET + CoreConstants.SEE_FNP_NOT_SET);
+    }
+
+    if (maxIndex < minIndex) {
+      LOG.warn("MaxIndex {} cannot be smaller than MinIndex {}.", maxIndex, minIndex);
+      maxIndex = minIndex;
+    }
+
+    if ((maxIndex - minIndex) > MAX_WINDOW_SIZE) {
+      LOG.warn("Large window sizes are not allowed.");
+      maxIndex = minIndex + MAX_WINDOW_SIZE;
+      LOG.warn("MaxIndex reduced to " + maxIndex);
+    }
+
+    IntegerTokenConverter itc = fileNamePattern.getIntegerTokenConverter();
+
+    if (itc == null) {
+      throw new IllegalStateException("FileNamePattern ["
+                                        + fileNamePattern.getPattern()
+                                        + "] does not contain a valid IntegerToken");
+    }
+    processedIndex = maxIndex;
+    super.start();
+  }
+
+  @Override
+  public void rollover() throws RolloverFailure {
+    Location parentLocation = getParent(activeFileLocation);
+
+    if (parentLocation == null) {
+      return;
+    }
+
+    // If maxIndex <= 0, then there is no file renaming to be done.
+    if (maxIndex >= 0) {
+      try {
+        String fileName = fileNamePattern.convertInt(maxIndex);
+        // Delete the oldest file. If processedIndex is not pointing to maxIndex, that means there was some
+        // exception while doing rename of files in previous rollover. So do not delete the existing file with
+        // maximum index otherwise we will end up deleting rolled over file
+        if (processedIndex == maxIndex) {
+          Location deleteLocation = parentLocation.append(fileName);
+          // no need to proceed further if we are not able to delete location so throw exception
+          if (deleteLocation.exists() && !deleteLocation.delete()) {
+            LOG.warn("Failed to delete location: {}", deleteLocation.toURI().toString());
+            throw new RolloverFailure(String.format("Not able to delete file: %s", deleteLocation.toURI().toString()));
+          }
+          processedIndex--;
+        }
+
+        for (int i = processedIndex; i >= minIndex; i--) {
+          String toRenameStr = fileNamePattern.convertInt(i);
+          Location toRename = parentLocation.append(toRenameStr);
+
+          // no point in trying to rename an non existent file
+          if (toRename.exists()) {
+            Location newName = parentLocation.append(fileNamePattern.convertInt(i + 1));
+            // throw exception if rename fails, so that in next iteration of rollover, it will be retried
+            if (toRename.renameTo(newName) == null) {
```

Comment:
this could throw `IOException` too, we need to update processedIndex also in that case.